### PR TITLE
Add focalnet backbone

### PR DESCRIPTION
**Pull request by alaradirik**
# What does this PR do?
Adds `FocalNetBackbone` class to be used by X-Decoder and possibly other frameworks as FocalNet was published fairly recently.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X ] Did you write any new necessary tests?

### CODE REVIEW

File: tests/models/focalnet/test_modeling_focalnet.py [hunk 1]
```diff
@@ -139,6 +148,28 @@ def create_and_check_model(self, config, pixel_values, labels):
...
+        # verify backbone works with out_features=None
+        config.out_features = None
+        model = FocalNetBackbone(config=config)
+        model.to(torch_device)
+        model.eval()
+        result = model(pixel_values)
+
+        # verify feature maps
+        self.parent.assertEqual(len(result.feature_maps), 1)
+        print(result.feature_maps[0].shape)
```
Comment (sgugger) [inline tests/models/focalnet/test_modeling_focalnet.py:170]:
```suggestion
```
To remove

**Review by sgugger — APPROVED** [review]
LGTM, but could you also check everything looks good @amyeroberts since you are more familiar with the backbones.

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for adding. 

Just a comment on a test that should be rewritten before merging

File: tests/test_backbone_common.py [hunk 2]
```diff
@@ -135,6 +135,8 @@ def test_backbone_common_attributes(self):
             # Verify num_features has been initialized in the backbone init
             self.assertIsNotNone(backbone.num_features)
             self.assertTrue(len(backbone.channels) == len(backbone.out_indices))
+            print(backbone.stage_names)
+            print(backbone.num_features)
```
Comment (amyeroberts) [inline tests/test_backbone_common.py:139]: @alaradirik  Can you create a PR to remove these?

**Pull request merged**